### PR TITLE
[Sana Sprint] add image-to-image pipeline 

### PR DESCRIPTION
**Pull request by linoytsaban**
This PR adds sana sprint image-to-image ⚡️⚡️
@sayakpaul @apolinario 

```py
import torch
from diffusers import SanaSprintImg2ImgPipeline
from diffusers.utils.loading_utils import load_image

image = load_image(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/penguin.png"
)

pipe = SanaSprintImg2ImgPipeline.from_pretrained("Efficient-Large-Model/Sana_Sprint_1.6B_1024px_diffusers", torch_dtype=torch.bfloat16)
pipe.to("cuda")

pipe(prompt="a cute pink bear", 
        image=image, 
        strength=0.45, 
        height=832, width=480).images[0]
```
run time on H100 - 
```
100%
 1/1 [00:00<00:00, 14.24it/s]
```
download (45)
download (47)

### CODE REVIEW

**Comment by linoytsaban** [comment]
@bot /style

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py [hunk 1]
```diff
@@ -0,0 +1,963 @@
...
+        ```py
+        >>> import torch
+        >>> from diffusers import SanaSprintImg2ImgPipeline
+
+        >>> pipe = SanaSprintImg2ImgPipeline.from_pretrained(
+        ...     "Efficient-Large-Model/Sana_Sprint_1.6B_1024px_diffusers", torch_dtype=torch.bfloat16
+        ... )
+        >>> pipe.to("cuda")
+
+        >>> image = pipe(prompt="a tiny astronaut hatching from an egg on the moon")[0]
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:73]: Needs to be supplemented an image?

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py [hunk 2]
```diff
@@ -0,0 +1,963 @@
...
+
+        prompt_attention_mask = text_inputs.attention_mask
+        prompt_attention_mask = prompt_attention_mask.to(device)
+
+        prompt_embeds = self.text_encoder(text_input_ids.to(device), attention_mask=prompt_attention_mask)
+        prompt_embeds = prompt_embeds[0].to(dtype=dtype, device=device)
+
+        return prompt_embeds, prompt_attention_mask
+
+    def encode_prompt(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:261]: Copy statement?

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+        init_timestep = min(num_inference_steps * strength, num_inference_steps)
+
+        t_start = int(max(num_inference_steps - init_timestep, 0))
+        timesteps = self.scheduler.timesteps[t_start * self.scheduler.order :]
+        if hasattr(self.scheduler, "set_begin_index"):
+            self.scheduler.set_begin_index(t_start * self.scheduler.order)
+
+        return timesteps, num_inference_steps - t_start
+
+    def check_inputs(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:376]: Do we need to check anything on the input image?

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py [hunk 4]
```diff
@@ -0,0 +1,963 @@
...
+                scm_timestep = torch.sin(timestep) / (torch.cos(timestep) + torch.sin(timestep))
+
+                scm_timestep_expanded = scm_timestep.view(-1, 1, 1, 1)
+                latent_model_input = latents_model_input * torch.sqrt(
+                    scm_timestep_expanded**2 + (1 - scm_timestep_expanded) ** 2
+                )
+
+                # predict noise model_output
+                noise_pred = self.transformer(
+                    latent_model_input.to(dtype=transformer_dtype),
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:902]: Shoulf we not scale the input with the standard deviation? 
https://github.com/huggingface/diffusers/blob/c36f8487df35895421c15f351c7d360bd6800e5c/examples/research_projects/sana/train_sana_sprint_diffusers.py#L1429

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py [hunk 5]
```diff
@@ -0,0 +1,963 @@
...
+
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
+            )
+
+        # adapt from https://github.com/huggingface/diffusers/blob/c36f8487df35895421c15f351c7d360bd680[…]/examples/research_projects/sana/train_sana_sprint_diffusers.py
+        noise = randn_tensor(shape, generator=generator, device=device, dtype=dtype) * self.scheduler.config.sigma_data
+        # latents = self.scheduler.add_noise(image_latents, timestep, noise)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:629]: Remove?

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py [hunk 6]
```diff
@@ -0,0 +1,963 @@
...
+            height,
+            width,
+            torch.float32,
+            device,
+            generator,
+            latents,
+        )
+
+        # I think this is redundant given the scaling in prepare_latents
+        # latents = latents * self.scheduler.config.sigma_data
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:869]: After encoding with the VAE we do need something scaling though:
https://github.com/huggingface/diffusers/blob/c36f8487df35895421c15f351c7d360bd6800e5c/examples/research_projects/sana/train_sana_sprint_diffusers.py#L1357

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py (hunk 4 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:902]: Nevermind, it's already there in `prepare_latents()`.

**Review by sayakpaul — COMMENTED** [review]
Thanks! Left some comments. Also, I think we need to add to docs.

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py (hunk 6 above)
Comment (linoytsaban) [reply src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:869]: I commented it because I think it's redundant since we have this in `prepare_latents`: 
https://github.com/huggingface/diffusers/blob/ac4a132ef95f6550c4f65666755b62e6ad4e7239/src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py#L607
and:
https://github.com/huggingface/diffusers/blob/ac4a132ef95f6550c4f65666755b62e6ad4e7239/src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py#L628-L630

File: src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py (hunk 3 above)
Comment (linoytsaban) [reply src/diffusers/pipelines/sana/pipeline_sana_sprint_img2img.py:376]: hmm we don't do any specific checks on it in other img2img pipelines (sdxl, sd3, flux)

**Comment by sayakpaul** [comment]
Cc: @asomoza for testing (if you want to).

**Review by sayakpaul — APPROVED** [review]
Good to merge after adding a test.

**Comment by linoytsaban** [comment]
@bot /style

**Comment by linoytsaban** [comment]
@bot /style

**Comment by linoytsaban** [comment]
@bot /style

**Comment by linoytsaban** [comment]
@bot /style

**Comment by linoytsaban** [comment]
@bot /style
